front/main/app/components/discussion-categories.js

```
@@ -0,0 +1,187 @@
+		},
+
+		disabledObserver: Ember.observer('disabled', function () {
+			if (this.get('disabled') === true) {
+				this.send('selectAllCategory', false);
+			}
+		}),
+
+		isAllCategoriesSelected: Ember.computed('categories.@each.selected', function () {
+			return this.get('categories').isEvery('selected', false);
+		}),
+
+		localCategories: Ember.computed('categories.@each.selected', function () {
+			const categories = this.get('categories'),
+				localCategories = new Ember.A();
+
+			categories.forEach((category) => {
+				localCategories.pushObject(Ember.Object.create({
+					category,
+					collapsed: false,
+					selected: category.selected
+				}));
+			});
+
+			if (this.get('canShowMore')) {
+				this.collapseCategoriesAboveLimit(localCategories);
+			}
+
+			return localCategories;
+		}),
+
+		allCategorySelected: Ember.computed.oneWay('isAllCategoriesSelected'),
+
+		categoriesInputIdPrefix: Ember.computed('inputIdPrefix', function () {
+			return `${this.get('inputIdPrefix')}-discussion-category-`;
+		}),
+
+		/**
+		 * Track click on category
+		 * @param {boolean} isAllCategories
+		 *
+		 * @returns {void}
+		 */
+		trackCategory(isAllCategories) {
+			track(isAllCategories ? trackActions.AllCategoriesTapped : trackActions.CategoryTapped);
+		},
+
+		/**
+		 * @param {Ember.Array} localCategories
+		 *
+		 * @returns {void}
+		 */
+		setAllCategorySelected(localCategories) {
+			const isNothingSelected = localCategories.isEvery('selected', false),
+				allCategorySelected = this.get('allCategorySelected');
+
+			if (!allCategorySelected && isNothingSelected) {
+				this.set('allCategorySelected', true);
+			} else if (allCategorySelected && !isNothingSelected) {
+				this.set('allCategorySelected', false);
+			}
+		},
+
+		collapseCategoriesAboveLimit(localCategories) {
+			const visibleCategoriesCount = this.get('visibleCategoriesCount');
+
+			if (typeof visibleCategoriesCount === 'number') {
```

As we spoke offline, we will not introduce any changes to this code, but we will strive to find patterns that will ensure any programmers errors will be caught / thrown but not fail silently.